### ISSUE TITLE
Widevine Error 6008

### ISSUE THREAD

**Issue by tim-edgren** [shaka-project/shaka-player#4284]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?**
Yes

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
3.0.1 (cdn.jsdelivr.net/npm/shaka-player@3.0.1/dist/shaka-player.compiled.js)

**Please ask your question**
I'm getting this error from the browser associated with error code 6008: ""Failed to execute 'update' on 'MediaKeySession': Rejected with system code (89)"

I'm attempting to use a proxy to Comcast/MPX's entitlement service to return the DRM license to the player, but I keep getting error code 6008 (with the underlying system code 89) whenever I do so.  My proxy returns JSON like this:

```json
{"license": "CAUSxwUKwQIIAxIQFwW5F8wSBIaLBjM6L {much_of_license_removed} HpfQ=" } 
```

which I attempt to unwrap and decode as follows:

```js
    player.getNetworkingEngine().registerResponseFilter(function(type, response) {
      // Alias some utilities provided by the library.
      const StringUtils = shaka.util.StringUtils;
      const Uint8ArrayUtils = shaka.util.Uint8ArrayUtils;

      // Only manipulate license responses:
      if (type == shaka.net.NetworkingEngine.RequestType.LICENSE) {
        // This is the wrapped license, which is a JSON string.
        const wrappedString = StringUtils.fromUTF8(response.data);
        // Parse the JSON string into an object.
        const wrapped = JSON.parse(wrappedString);

        // This is a base64-encoded version of the raw license.
        const rawLicenseBase64 = wrapped.license;
        // Decode that base64 string into a Uint8Array and replace the response
        // data.  The raw license will be fed to the Widevine CDM.
        response.data = Uint8ArrayUtils.fromBase64(rawLicenseBase64);
        console.log(rawLicenseBase64);
      }
    });
```

Any help on deciphering that system code 89 (which I know is from Chrome, and seems to suggest that the license is malformed) would be greatly appreciated!

[shaka_error_6008.txt](https://github.com/shaka-project/shaka-player/files/8874382/shaka_error_6008.txt)

**Comment by tim-edgren** [comment]
OK, I've now learned/realized that the 'license' response I'm seeing is actually the server certificate, and that the second Widevine request is what is failing.  So I need to figure out:

- how to apply the retrieved server certificate to my Widevine CDM session
- how to format the second actual license request and allow it to pass through my proxy

Any help in this would be greatly appreciated.

**Comment by joeyparrish** [comment]
You can configure Shaka with the server certificate in advance.  You can give a URI, and the player will fetch it for you and feed it to the CDM in advance of creating a session:

```js
player.configure('drm.advanced.com\\.widevine\\.alpha.serverCertificateUri', 'https://hostname/path/to/servercert');
```

Or you can give us the bytes:

```js
player.configure('drm.advanced.com\\.widevine\\.alpha.serverCertificate', new Uint8Array(...));
```

Either way, we'll feed this to the CDM through EME's `setServerCertificate()` method, and the CDM won't send a server certificate request before the license request.

Does this help?

**Comment by joeyparrish** [comment]
As for the formatting of the license exchange, unless you inject request/response filters, everything that the CDM generates in a request (`message` event on `MediaKeySession`) is sent directly to your configured license server in a POST, and everything that server replies with is sent directly to the CDM in an `update()` call on `MediaKeySession`.

If there is any formatting your service uses on either request or response that wraps the raw Widevine protocol, you must wrap/unwrap those requests/responses in your application's filters.  See also our License Wrapping tutorial: https://shaka-player-demo.appspot.com/docs/api/tutorial-license-wrapping.html

I hope this helps!

**Comment by joeyparrish** [comment]
I don't understand where you go the base64 string.  Is this coming from a third-party service?  Unfortunately, we can't explain to you how that service is encoding or formatting things, since that's completely outside of our control.

Are there docs from that service that we could help you interpret?  Could you or the implementors of your license service reach out to Widevine for support?

**Comment by joeyparrish** [comment]
When the Widevine CDM sends the two-byte request (base64 as "CAQ"), it's requesting a service cert.  This is an extra round trip you can avoid by providing a service cert to Shaka Player in advance.  (We pass it on to the CDM before creating a session.)

What might surprise you is that the response to that CAQ request is not in exactly the same format that the CDM expects.  When this feature first launched, I tried to get Widevine to change it so that the CDM would accept both formats (one is a wrapped version of the other), but I was not successful in changing anyone's mind.  I still think this is confusing and makes it more difficult to use this feature.

So it's entirely possible that what Comcast sent you in base64 for the service cert is in the wrong format.  Hard to say for sure, though.

Do you have access to Widevine's integration console?  There is a "Diagnostics" panel there where you can ask it to parse a base64-encoded license response.  That might help you determine if the base64 service cert from Comcast is in the correct format or not.  If you don't have access, you can share the full base64 string with me and I'll check it for you.

**Comment by tim-edgren** [comment]
Yes, if I can, I definitely want to get out of the business of the two-phased Widevine round trip, if only so I can get a clearer picture of the real problem.  

I **do** have access to the Widevine integration platform -- I tested the base64 'certificate' I got from Comcast there under the Diagnostic Tools section.  It generates an Error if I process it as a _response_, but if I send it as a **license request**, it gives me back a SERVICE_CERTIFICATE message.  I'm not sure what that means -- is it possible that what I have is actually a license request, and not a certificate at all?

**Comment by tim-edgren** [comment]
I'm really struggling to articulate the problem I'm still having.  Maybe I can sketch out the sequence of events, as I see them from the console and network tabs:

1) the player loads the manifest MPD and sees that the content I'm trying to play has widevine encryption
2) the call to the license server (with the releasePid and the "CAQ" challenge) is initiated
3) the Comcast license server responds with what I think is a valid license request (because when I run it through the widevine diagnostic tools, it responds with a SERVICE_CERTIFICATE msg)
4) I process the response with the following code to make it useable by the Shaka player:
`      if (type == shaka.net.NetworkingEngine.RequestType.LICENSE) {
        // This is the wrapped license, which is a JSON string.
        const wrappedString = StringUtils.fromUTF8(response.data);
        // Parse the JSON string into an object.
        const wrapped = JSON.parse(wrappedString);
        // This is a base64-encoded version of the raw license.
        var rawLicenseBase64 = wrapped.getWidevineLicenseResponse.license;
        // Decode that base64 string into a Uint8Array and replace the response data.  
        //The raw license will be fed to the Widevine CDM.
        var binary_string = window.atob(rawLicenseBase64);
        var len = binary_string.length;
        var bytes = new Uint8Array(len);
        for (var i=0; i < len; i++) {
          bytes[i] = binary_string.charCodeAt(i);
        }
        response.data = bytes.buffer;
        return response.data;
      }
`
5) the CDM somehow uses that certificate response to come up with an actual license request, which is put into the payload
6) I intercept that request and reformat it into JSON so the Comcast server can see it:
`      if (type == shaka.net.NetworkingEngine.RequestType.LICENSE) {
        // Create the wrapped request structure.
        const wrapped = { };
        wrapped.getWidevineLicense = { };
        // Encode the raw license request in base64.
        wrapped.getWidevineLicense.widevineChallenge = Uint8ArrayUtils.toBase64(new Uint8Array(request.body));
        wrapped.getWidevineLicense.releasePid = releasePid;
        // Encode the wrapped request as JSON.
        const wrappedJson = JSON.stringify(wrapped);
        // Convert the JSON string back into an ArrayBuffer to replace the request body.
        request.body = StringUtils.toUTF8(wrappedJson);
        return request.body;
      }
`
7) the Comcast server says it is not a valid license challenge:
{"title":"com.theplatform.module.exception.ValidationException","description":"Widevine generateLicense call against [https://license.widevine.com/cenc/getlicense/theplatformpureflix] failed with status [INVALID_LICENSE_CHALLENGE]: Payload should be non-empty.","responseCode":422,"correlationId":"c42c9016-b858-4093-a316-84e9aaa61483","isException":true}
(I've noticed that the 'payload should be non-empty' error is produced even when the payload is properly formatted but has the wrong challenge.)

I'm baffled as to what I can do to resolve this.

**Comment by tim-edgren** [comment]
Got some troubleshooting help from David Eisenbacher (EZDRM) who correctly identified that my problem had to do with the base64 encoding between the Shaka player and my widevine license server.  The root issue seems to be that Shaka's Uint8ArrayUtils.toBase64 routine expects the RFC 4648 base64url encoding, while Comcast is using RFC 4648 base64 (standard).  Once I added code to manually translate the 62nd and 63rd characters (+ and /) to the characters Shaka expects (- and _), I was able to get past the INVALID_LICENSE_CHALLENGE issue.  

Of course, now I'm seeing a 4012 error having to do with missing keys, but I will open that up as a separate issue.
